sql/api/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -117,65 +130,144 @@ object JavaTypeInference {
+          tv.getBounds
+            .flatMap(getAllSuperClasses)
+            .foldLeft[(Option[AgnosticEncoder[_]], Option[AgnosticEncoder[_]])](None -> None) {
+              case (r @ (Some(_), _), _) => r
 
-    case pt: ParameterizedType =>
+              case (r @ (None, serEncoder), bound) =>
+                Try(encoderFor(bound, seenTypeSet, typeVariables, forGenericBound = true)) match {
+                  case Success(value) => (Option(value), serEncoder)
+
+                  case Failure(UseSerializationEncoder(encoderProvider)) if serEncoder.isEmpty =>
+                    None -> Option(encoderProvider(concreteBound.getOrElse(bound)))
+
+                  case Failure(_) => r
+                }
+            } match {
+            case (Some(encd), _) => encd
+
+            case (None, Some(encd)) => encd
+
+            case _ => throw ExecutionErrors.cannotFindEncoderForTypeError(t.getTypeName)
+          }
+      }
+
+    case pt: ParameterizedType if !forGenericBound =>
       encoderFor(pt.getRawType, seenTypeSet, JavaTypeUtils.getTypeArguments(pt).asScala.toMap)
 
-    case c: Class[_] =>
+    case c: Class[_] if !forGenericBound =>
       if (seenTypeSet.contains(c)) {
         throw ExecutionErrors.cannotHaveCircularReferencesInBeanClassError(c)
       }
 
-      // TODO: we should only collect properties that have getter and setter. However, some tests
-      //   pass in scala case class as java bean class which doesn't have getter and setter.
+      // TODO: we should only collect properties that have getter and setter. However, some
+      //  tests pass in scala case class as java bean class which doesn't have getter and
+      //  setter.
       val properties = getJavaBeanReadableProperties(c)
-      // add type variables from inheritance hierarchy of the class
-      val classTV = JavaTypeUtils.getTypeArguments(c, classOf[Object]).asScala.toMap ++
-        typeVariables
-      // Note that the fields are ordered by name.
-      val fields = properties.map { property =>
-        val readMethod = property.getReadMethod
-        val encoder = encoderFor(readMethod.getGenericReturnType, seenTypeSet + c, classTV)
-        // The existence of `javax.annotation.Nonnull`, means this field is not nullable.
-        val hasNonNull = readMethod.isAnnotationPresent(classOf[Nonnull])
-        EncoderField(
-          property.getName,
-          encoder,
-          encoder.nullable && !hasNonNull,
-          Metadata.empty,
-          Option(readMethod.getName),
-          Option(property.getWriteMethod).map(_.getName))
+      // if the properties is empty and this is not a top level enclosing class, then we
+      // should not consider class as bean, as otherwise it will be treated as empty schema
+      // and loose the data on deser.
+      if (properties.isEmpty && seenTypeSet.nonEmpty) {
+        if (classOf[KryoSerializable].isAssignableFrom(c)) {
```

Comment:
This will be an issue for Connect. While the API supports Kryo, Connect can't support Kryo in its current form. Either we have detect whether we are in connect mode, or we have to just fall back to java serialization.

Reply:
sure. I was not aware of that.. will write test for it and check.
will be checking in a code with some refactoring..

Reply:
@hvanhovell Added check for client connect while creating encoders, to not use kryo based encoder. Using threadlocal to detect the same, instead of api change.